File: threadapp/src/main/java/com/example/threadapp/controller/ThreadController.java
package com.example.threadapp.controller;

import com.example.threadapp.model.Thread;
import com.example.threadapp.service.ThreadService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/threads")
public class ThreadController {

    private final ThreadService threadService;

    @Autowired
    public ThreadController(ThreadService threadService) {
        this.threadService = threadService;
    }

    @GetMapping
    public List<Thread> getAllThreads() {
        return threadService.getAllThreads();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Thread> getThreadById(@PathVariable String id) {
        return threadService.getThreadById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping
    public ResponseEntity<Thread> createThread(@RequestBody Thread thread) {
        Thread createdThread = threadService.createThread(thread);
        return new ResponseEntity<>(createdThread, HttpStatus.CREATED);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Thread> updateThread(@PathVariable String id, @RequestBody Thread threadDetails) {
        return threadService.updateThread(id, threadDetails)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Thread> deleteThread(@PathVariable String id) {
        return threadService.getThreadById(id)
                .map(thread -> {
                    threadService.deleteThread(id);
                    return ResponseEntity.ok(thread);
                })
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/author/{authorId}")
    public List<Thread> getThreadsByAuthorId(@PathVariable String authorId) {
        return threadService.getThreadsByAuthorId(authorId);
    }

    @GetMapping("/sorted/upvotes")
    public List<Thread> getThreadsSortedByUpvotes() {
        return threadService.getThreadsSortedByUpvotes();
    }

    @GetMapping("/sorted/downvotes")
    public List<Thread> getThreadsSortedByDownvotes() {
        return threadService.getThreadsSortedByDownvotes();
    }

    @GetMapping("/sorted/createdAt")
    public List<Thread> getThreadsSortedByCreatedAt() {
        return threadService.getThreadsSortedByCreatedAt();
    }

    @GetMapping("/sorted/createdAtAsc")
    public List<Thread> getThreadsSortedByCreatedAtAsc() {
        return threadService.getThreadsSortedByCreatedAtAsc();
    }
}

File: ThreadApp/src/controllers/threadController.py
package com.wroteit.ThreadApp.controller;

import com.wroteit.ThreadApp.model.ThreadPost;
import com.wroteit.ThreadApp.repository.ThreadPostRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.List;

@RestController
@RequestMapping("/api/threads")
@RequiredArgsConstructor
public class ThreadController {

    private final ThreadPostRepository threadRepo;

    @PostMapping
    public ResponseEntity<ThreadPost> createThread(@RequestBody ThreadPost thread) {
        thread.setCreatedAt(LocalDateTime.now());
        ThreadPost saved = threadRepo.save(thread);
        return ResponseEntity.ok(saved);
    }

    @GetMapping
    public ResponseEntity<List<ThreadPost>> getAllThreads() {
        return ResponseEntity.ok(threadRepo.findAll());
    }

    @GetMapping("/{id}")
    public ResponseEntity<ThreadPost> getThreadById(@PathVariable Long id) {
        return threadRepo.findById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteThread(@PathVariable Long id) {
        if (threadRepo.existsById(id)) {
            threadRepo.deleteById(id);
            return ResponseEntity.noContent().build();
        }
        return ResponseEntity.notFound().build();
    }
}
